keps/sig-node/2837-pod-level-resource-spec/README.md

```diff
@@ -0,0 +1,1806 @@
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+Currently resource allocation on PodSpec is container-centric, allowing users to
+specify resource requests and limits for each container. The scheduler uses the
+aggregate of the requested resources by all containers in a pod to find a
+suitable node for scheduling the pod. The kubelet then enforces these resource
+constraints by translating the requests and limits into corresponding cgroup
+settings both for containers and for the pod (where pod level values are
+aggregates of container level values). The existing Pod API lacks the ability to
+set resource constraints at pod level, limiting the flexibility and ease of
+resource management for pods as a whole. This limitation is particularly
+problematic when users want to focus on  controlling the overall resource
+consumption of a pod without the need to meticulously configure resource
+specifications for each individual container in it.
+
+To address this, this KEP proposes extending the Pod API to support Pod-level
+resource limits and requests for non-extended resources (CPU and Memory), in
+addition to the existing container-level settings.
+
+## Motivation
+
+Kubernetes workloads often consist of multiple containers collaborating within a
+pod. While this separation enables granular monitoring and management of each
+container, it can be challenging to accurately estimate and allocate resources
+for individual containers, especially for workloads with unpredictable or bursty
+resource demands. This often leads to over-allocation of resources to ensure that
+no container experiences resource starvation, as kubernetes currently lacks a
+mechanism for sharing resources between containers easily. 
+
+Introducing pod-level resource requests and limits offers a more simplified
+resource management for multi-container pods as it is easier to gauge theß
+collective resource usage of all containers in a pod rather than predicting each
+container’s individual needs. This proposal seeks to support pod-level resource
+management, enabling Kubernetes to control the total resource consumption of the
+pod, relieving users from the burden of meticulously configuring resources for
+each container. This simplified approach is particularly valuable for tightly
+coupled applications, and it will enable Kubernetes to leverage the underlying
+cgroup resource management mechanisms more effectively for better resource
+utilization for bursty workloads. 
+
+This new feature of resource specification at pod-level would complement, not
+replace, the existing container-level limits, providing users with greater
+flexibility.
+
+### Goals
+
+This proposal aims to:
+
+1. Extend Pod API to allow specifying resource spec (requests and limits) at the
+   pod level. 
+2. Make this feature compatible with existing usage of container level
+   requests/limits, and other features like topology manager, memory manager,
+   etc.
+
+### Non Goals
```

Comment:
Can we explicitly add the following items into Non Goals: 

1) Pod-level device plugins: existing container-level device plugins are compatible with this proposal, but there are some discussion on how to share device(s) among containers within a pod. Treat those as none goal explicitly.
2) Dynamic QoS class adjustments: As we discussed offline, we agreed a static QoS class determination based on pod-level and container-level resources. Please explictly state that exploring dynamic QoS changes based on runtime conditions or pod phases is Non Goal. We can explore the possibility in the future. 
3) Fine-grained control over resource sharing within a pod: While pod-level resources enable basic resource sharing, more advanced controls (e.g., weighted sharing, prioritization) could be considered as a future enhancement.
4) Performance optimizations for specific workloads: While the feature is expected to improve resource utilization, specific performance optimizations for certain workloads (e.g., high-performance computing) can be explored in future iterations.